ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+                    * `tail(x[i]) = ""` (the empty string)
+                * Otherwise:
+                    * `head(x[i]) = ""` (the empty string)
+                    * `tail(x[i]) = ""` (the empty string)
+            * Otherwise:
+                * `head(x[i]) = enc(x[i])`
+                * `tail(x[i]) = ""` (the empty string)
+        * Otherwise:
+            * `head(x[i]) = enc(len( head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[i-1]) ))`
+            * `tail(x[i]) = enc(x[i])`
+* If `x` is a fixed-length array `T[N]`:
+    * `enc(x) = enc((x[0], ..., x[N-1]))`, i.e. it’s encoded as if it were an `N` element tuple where every element is type `T`.
+* If `x` is a variable-length array `T[]` with `k` elements:
+    * `enc(x) = enc(k) enc([x[0], ..., x[k-1]])`, i.e. it’s encoded as if it were a fixed-length array of `k` elements, prefixed with its length, `k`.
+* If `x` is an `N`-bit unsigned integer, `uint<N>`:
+    * `enc(x)` is the `N`-bit big-endian encoding of `x`.
+* If `x` is an `N`-bit unsigned fixed-point decimal number with precision `M`, `ufixed<N>x<M>`:
+    * `enc(x) = enc(x * 10^M)`, where `x * 10^M` is interpreted as a `uint<N>`.
+* If `x` is a boolean value `bool`:
+    * `enc(x)` is a single byte whose **most significant bit** is either 1 or 0, if `x` is true or false respectively. All other bits are 0. Note: this means that a value of true will be encoded as `0x80` (`10000000` in binary) and a value of false will be encoded as `0x00`. This is in contrast to most other encoding schemes, where a value of true is encoded as `0x01`.
+
+Since `string` and `address` are aliases for `byte[]` and `byte[32]`
+respectively (and `byte` is an alias for `uint8`), the rules for
+encoding these types are covered above.
+
+## Further considerations
+
+This document describes argument and return value encoding only.
+There are several other aspects to invoking an Algorand app properly
+that require standardization to improve high-level access to
+contracts.
+
+### Opt In, Close Out, and Clear State.
+
+An ARC-4 app **MUST** allow any account to opt-in and close-out, using
+an application call transaction with no arguments an OnComplete equal
+to . Clients **MAY** determine that such an opt-in is unnecessary by
+querying the blockchain for the app's schema. Opt-in is unnessary if
+the app has no local schema.  However, even if a local schema is
+present, some methods may be callable without opt-in. Clients **MAY**
+attempt method calls without opt-in.
+
+Clients **MAY** display the minimum balance increment that such an
+opt-in would require.  If an opt-in or close-out contains arguments,
+an ARC-4 app **MUST** fail the transaction, or interpret the arguments
+as an invocation as described in the rest of this document, to perform
+along with the opt-in or close-out.
```

Comment:
This sentence is sensible, but seems incompatible with the discussion above

Reply:
I think we need to make it clear that this only works if _optIn has not been specified, which implies that an empty optin is allowed.

Reply:
Just to be clear, the way that Reach works is that *any* method can have an `OptIn`, unless it must have `DeleteApplication` AND there is also a "nop" method that must have an `OptIn`.